### PR TITLE
New feature: r-apt

### LINKED ISSUE #66: New feature: r-apt
Set up R to install R packages via apt (with `bspm`).

ToDo: I don't know if bspm will work on arm64.

### PR DESCRIPTION
**Pull request by eitsupi**
Add new feature to install R via apt, close #66.

~~Ideally I would like to add the same options as r-rig, but it seems that I cannot install `languageserver` and `httpgd` via apt in Debian.
Perhaps the option to install them should be configured to not work with Debian...~~

Install languageserver and httpgd from CRAN.

### CODE REVIEW

File: src/r-apt/install.sh [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+# shellcheck source=/dev/null
+source /etc/os-release
+
+if grep -q "Ubuntu" </etc/os-release; then
+    check_packages curl ca-certificates
+    curl -fsSL https://cloud.r-project.org/bin/linux/ubuntu/marutter_pubkey.asc | tee -a /etc/apt/trusted.gpg.d/cran_ubuntu_key.asc
+    echo "deb [arch=amd64] https://cloud.r-project.org/bin/linux/ubuntu ${UBUNTU_CODENAME}-cran40/" >/etc/apt/sources.list.d/cran-ubuntu.list
+    curl -fsSL https://eddelbuettel.github.io/r2u/assets/dirk_eddelbuettel_key.asc | tee -a /etc/apt/trusted.gpg.d/cranapt_key.asc
+    echo "deb [arch=amd64] https://dirk.eddelbuettel.com/cranapt ${UBUNTU_CODENAME} main" >/etc/apt/sources.list.d/cranapt.list
```
Comment (eddelbuettel) [inline src/r-apt/install.sh:63]: Please (generally) use https://r2u.stat.illinois.edu/ubuntu -- it has much better connectivity.

**Review by eddelbuettel — APPROVED** [review]

File: src/r-apt/install.sh (hunk 1 above)
Comment (eddelbuettel) [reply src/r-apt/install.sh:63]: >  I cannot install languageserver and httpgd via apt

If you want them badly on Debian, maybe create a small repo that hosts binaries you provide?

**Comment by eitsupi** [comment]
> If you want them on Debian, one way is to create a small repo (you provide somewhere) which hosts binaries (you have to make ) ? Might be overkill if a one-time source installation works?

You are right, source installation should be sufficient for these.
Some dependencies should be able to be installed with apt, so check which packages can be installed with apt.

File: src/r-apt/install.sh (hunk 1 above)
Comment (eitsupi) [reply src/r-apt/install.sh]: Thanks for your review. Done.

**Review by eddelbuettel — APPROVED** [review]
Also, we're small: "the" mirror, as in "the main mirror" or the "the sole mirror" (I know of, anyway...)

**Comment by eitsupi** [comment]
I wondered what was wrong because the installation failed no matter how many times I tried and this was the problem...
https://github.com/ManuelHentschel/vscDebugger/pull/173

**Comment by eddelbuettel** [comment]
R itself definitely can have utf-8 in DESCRIPTION.  I know, because the static html 'compiler' I use for [CRANberries]() is so old that I explicitly filter those out as that converter from txt to html (and rss feed) can only do 7-bit as if it were still the 1990s....

Setting encoding may help.  But you of course may also have trouble coming from another angle here as there are more moving pieces.

**Comment by eitsupi** [comment]
> I wondered what was wrong because the installation failed no matter how many times I tried and this was the problem...
> [ManuelHentschel/vscDebugger#173](https://github.com/ManuelHentschel/vscDebugger/pull/173)

Using `pak` instead of `remotes` has worked around this problem.

Now debian is able to complete the installation successfully, but the next problem is with `ubuntu:focal`......

@eddelbuettel Do you know what is causing this error, it is occurring in focal?

```log
 The following packages have unmet dependencies:
  r-cran-devtools : Depends: r-cran-miniui but it is not going to be installed
                    Depends: r-cran-pkgdown but it is not going to be installed
                    Depends: r-cran-profvis but it is not going to be installed
                    Depends: r-cran-testthat but it is not going to be installed
  r-cran-irkernel : Depends: r-cran-repr but it is not going to be installed
                    Depends: r-cran-irdisplay but it is not going to be installed
                    Depends: r-cran-pbdzmq but it is not going to be installed
  r-cran-rmarkdown : Depends: r-cran-bslib but it is not going to be installed
                     Depends: r-cran-htmltools but it is not going to be installed
                     Depends: r-cran-jquerylib but it is not going to be installed
```

https://github.com/rocker-org/devcontainer-features/actions/runs/3256350662/jobs/5346702229#step:4:4695

**Comment by eddelbuettel** [comment]
Looks like you didn't set pinning up -- then the in distr packages may (accidentally) sort higher.  

Quoting from my README.md:


### Pinning

Because we let `apt` (and related tools) pick the packages, we have to ensure
that the CRANapt repo sorts higher than the default repo as (older)
package builds in the distribution itself may appear (to `apt`) to be
newer. A case in point was package `gtable` whose version in Ubuntu was
`0.3.0+dfsg-1` which accidentally sorts higher than the rebuild we made under
a newer and more consistent version number `0.3.0-1.ca2004.1`.  One possible
fix is 'apt pinning'. Place a file `/etc/apt/preferences.d/99cranapt` with content

    Package: *
    Pin: release o=CRAN-Apt Project
    Pin: release l=CRAN-Apt Packages
    Pin-Priority: 700

which will now give packages from this repo a higher default priority of 700
overriding the standard value of 500.

**Comment by eitsupi** [comment]
> Looks like you didn't set pinning up -- then the in distr packages may (accidentally) sort higher.

I see, thank for pointing this out!

**Comment by eitsupi** [comment]
It works now 🎉

**Comment by eddelbuettel** [comment]
> I think the documentation needs a little more improvement.

I am all ears. Here (PRs) or over slack.  But the 'pinning' is a) in the README, b) in the setup scripts and c) in the Docker containers. I am not the best at reading docs either but I tried :smile:

**Comment by eitsupi** [comment]
@eddelbuettel Sorry, the documentation I am referring to is the documentation for this feature, not r2u.
Your description was clear, and I simply skipped over it, and the setup was simple!

**Comment by eitsupi** [comment]
I have created an issue for an option to configure bspm that can be discussed later (#69).
For now this will be merged and released.

**Comment by eitsupi** [comment]
Now listed on the web site🎉

image
https://containers.dev/features

**Pull request merged**